core/src/main/java/org/infinispan/marshall/core/ClassToExternalizerMap.java

```diff
@@ -0,0 +1,447 @@
+
+import java.util.Arrays;
+
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * An efficient identity object map whose keys are Class objects and whose
+ * values are {@code int}s.
+ */
+final class ClassToExternalizerMap {
+
+   private static final Log log = LogFactory.getLog(ClassToExternalizerMap.class);
+
+   private AdvancedExternalizer[] values;
+   private Class[] keys;
+   private int count;
+   private int resizeCount;
+   private final float loadFactor;
+
+   /**
+    * Construct a new instance with the given initial capacity and load factor.
+    *
+    * @param initialCapacity the initial capacity
+    * @param loadF the load factor
+    */
+   public ClassToExternalizerMap(int initialCapacity, final float loadF) {
+      if (initialCapacity < 1) {
+         throw new IllegalArgumentException("initialCapacity must be > 0");
+      }
+      if (loadF <= 0.0f || loadF >= 1.0f) {
+         throw new IllegalArgumentException("loadFactor must be > 0.0 and < 1.0");
+      }
+      if (initialCapacity < 16) {
+         initialCapacity = 16;
+      } else {
+         // round up
+         final int c = Integer.highestOneBit(initialCapacity) - 1;
+         initialCapacity = Integer.highestOneBit(initialCapacity + c);
+      }
+      keys = new Class[initialCapacity];
+      values = new AdvancedExternalizer[initialCapacity];
+      resizeCount = (int) ((double) initialCapacity * (double) loadF);
+      this.loadFactor = loadF;
+   }
+
+   /**
+    * Construct a new instance with the given load factor and an initial capacity of 64.
+    *
+    * @param loadFactor the load factor
+    */
+   public ClassToExternalizerMap(final float loadFactor) {
+      this(64, loadFactor);
+   }
+
+   /**
+    * Construct a new instance with the given initial capacity and a load factor of {@code 0.5}.
+    *
+    * @param initialCapacity the initial capacity
+    */
+   public ClassToExternalizerMap(final int initialCapacity) {
+      this(initialCapacity, 0.5f);
+   }
+
+   /**
+    * Construct a new instance with an initial capacity of 64 and a load factor of {@code 0.5}.
+    */
+   public ClassToExternalizerMap() {
+      this(0.5f);
+   }
+
+   /**
+    * Get a value from the map.
+    *
+    * @param key the key
+    * @return the map value at the given key, or null if it's not found
+    */
+   public AdvancedExternalizer get(Class key) {
+      final Class[] keys = this.keys;
+      final int mask = keys.length - 1;
+      int hc = System.identityHashCode(key) & mask;
+      Class k;
+      for (;;) {
+         k = keys[hc];
+         if (k == key) {
+            return values[hc];
+         }
+         if (k == null) {
+            // not found
+            return null;
+         }
+         hc = (hc + 1) & mask;
+      }
+   }
+
+   /**
+    * Put a value into the map.  Any previous mapping is discarded silently.
```

Comment:
Hmmm, could be. I've separate this to [ISPN-7210](https://issues.jboss.org/browse/ISPN-7210).